File: code/javascript/dynamic_programming_article-08.js
/**
 * Thinking transformation between dynamic programming and backtracking algorithms
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return backtrack(...args);
}

var wordBreak = function(s, wordDict) {
  let res = [];
  let track = [];
  function backtrack(s, i, wordDict) {
    if (i === s.length) {
      res.push(track.join(" "));
      return;
    }
    if (i > s.length) {
      return;
    }
    for (let word of wordDict) {
      let len = word.length;
      if (i + len > s.length) {
        continue;
      }
      let subStr = s.substring(i, i + len);
      if (subStr !== word) {
        continue;
      }
      track.push(word);
      backtrack(s, i + len, wordDict);
      track.pop();
    }
  }
  backtrack(s, 0, wordDict);
  return res;
};

File: code/python/dynamic_programming_article-08.py
"""Thinking transformation between dynamic programming and backtracking algorithms"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().__init__(*args, **kwargs)


class _Solution:
    def __init__(self):
        self.res = []
        self.track = []
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        self.backtrack(s, 0, wordDict)
        return self.res
    def backtrack(self, s: str, i: int, wordDict: List[str]) -> None:
        if i == len(s):
            self.res.append(" ".join(self.track))
            return
        if i > len(s):
            return
        for word in wordDict:
            length = len(word)
            if i + length > len(s):
                continue
            sub_str = s[i:i + length]
            if sub_str != word:
                continue
            self.track.append(word)
            self.backtrack(s, i + length, wordDict)
            self.track.pop()
